// @paid-ai/paid-node/dist/esm/integrations/adapters/nextjs/base-adapter.mjs
/**
 * Next.js adapter for Paid SDK
 * Converts Next.js App Router requests to framework-agnostic format
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { NextResponse } from "next/server";
import { createFrameworkAdapter } from "../../utils/create-adapter.mjs";
/**
 * Next.js adapter for App Router
 * Handles NextRequest/NextResponse conversion to base handler format
 */
export const nextjsAdapter = createFrameworkAdapter({
    getBody: (req) => __awaiter(void 0, void 0, void 0, function* () {
        if (req.method === "GET" || req.method === "HEAD") {
            return {};
        }
        try {
            return yield req.json();
        }
        catch (error) {
            return {};
        }
    }),
    getHeaders: (req) => {
        const headers = {};
        req.headers.forEach((value, key) => {
            headers[key] = value;
        });
        return headers;
    },
    getMethod: (req) => req.method || "GET",
    getParams: (_req, context) => __awaiter(void 0, void 0, void 0, function* () {
        if (!(context === null || context === void 0 ? void 0 : context.params)) {
            return {};
        }
        const params = context.params instanceof Promise ? yield context.params : context.params;
        return params || {};
    }),
    getQuery: (req) => {
        const query = {};
        req.nextUrl.searchParams.forEach((value, key) => {
            query[key] = value;
        });
        return query;
    },
    sendJson: (data, status) => NextResponse.json(data, { status }),
});
