// @paid-ai/paid-node/dist/esm/tracing/wrappers/openAiWrapper.mjs
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { SpanStatusCode } from "@opentelemetry/api";
import { getPaidTracer, getToken } from "tracing/tracing.js";
import { getTracingContext } from "tracing/tracingContext.js";
export class PaidOpenAI {
    constructor(openaiClient) {
        this.openai = openaiClient;
        const tracer = getPaidTracer();
        if (!tracer) {
            throw new Error("Paid tracer is not initialized, Make sure to call 'initializeTracing()' first");
        }
        this.tracer = tracer;
    }
    get chat() {
        return new ChatWrapper(this.openai, this.tracer);
    }
    get embeddings() {
        return new EmbeddingsWrapper(this.openai, this.tracer);
    }
    get images() {
        return new ImagesWrapper(this.openai, this.tracer);
    }
    get responses() {
        return new ResponsesWrapper(this.openai, this.tracer);
    }
}
class ChatWrapper {
    constructor(openai, tracer) {
        this.openai = openai;
        this.tracer = tracer;
    }
    get completions() {
        return new ChatCompletionsWrapper(this.openai, this.tracer);
    }
}
class ChatCompletionsWrapper {
    constructor(openai, tracer) {
        this.openai = openai;
        this.tracer = tracer;
    }
    create(params) {
        return __awaiter(this, void 0, void 0, function* () {
            const token = getToken();
            const { externalProductId, externalCustomerId } = getTracingContext();
            if (!token || !externalCustomerId) {
                throw new Error("No token or externalCustomerId: This wrapper should be used inside a callback to paid.trace().");
            }
            return this.tracer.startActiveSpan("trace.openai.chat", (span) => __awaiter(this, void 0, void 0, function* () {
                const attributes = {
                    "gen_ai.system": "openai",
                    "gen_ai.operation.name": "chat",
                };
                attributes["external_customer_id"] = externalCustomerId;
                attributes["token"] = token;
                if (externalProductId) {
                    attributes["external_agent_id"] = externalProductId;
                }
                span.setAttributes(attributes);
                try {
                    const response = (yield this.openai.chat.completions.create(params));
                    if (response.usage) {
                        span.setAttributes({
                            "gen_ai.usage.input_tokens": response.usage.prompt_tokens,
                            "gen_ai.usage.output_tokens": response.usage.completion_tokens,
                            "gen_ai.response.model": response.model,
                        });
                    }
                    span.setStatus({ code: SpanStatusCode.OK });
                    return response;
                }
                catch (error) {
                    span.setStatus({ code: SpanStatusCode.ERROR, message: error.message });
                    span.recordException(error);
                    throw error;
                }
                finally {
                    span.end();
                }
            }));
        });
    }
}
class ResponsesWrapper {
    constructor(openai, tracer) {
        this.openai = openai;
        this.tracer = tracer;
    }
    create(params) {
        return __awaiter(this, void 0, void 0, function* () {
            const { externalCustomerId, externalProductId } = getTracingContext();
            const token = getToken();
            if (!token || !externalCustomerId) {
                throw new Error("No token or externalCustomerId: This wrapper should be used inside a callback to paid.trace().");
            }
            return this.tracer.startActiveSpan("trace.openai.responses", (span) => __awaiter(this, void 0, void 0, function* () {
                var _a, _b;
                const attributes = {
                    "gen_ai.system": "openai",
                    "gen_ai.operation.name": "chat", // Equivalent to chat.completions
                };
                attributes["external_customer_id"] = externalCustomerId;
                attributes["token"] = token;
                if (externalProductId) {
                    attributes["external_agent_id"] = externalProductId;
                }
                span.setAttributes(attributes);
                try {
                    const response = yield this.openai.responses.create(params);
                    if (response.usage) {
                        span.setAttributes({
                            "gen_ai.usage.input_tokens": response.usage.input_tokens,
                            "gen_ai.usage.output_tokens": response.usage.output_tokens,
                            "gen_ai.response.model": response.model,
                        });
                        if ((_a = response.usage.input_tokens_details) === null || _a === void 0 ? void 0 : _a.cached_tokens) {
                            span.setAttribute("gen_ai.usage.cached_input_tokens", response.usage.input_tokens_details.cached_tokens);
                        }
                        if ((_b = response.usage.output_tokens_details) === null || _b === void 0 ? void 0 : _b.reasoning_tokens) {
                            span.setAttribute("gen_ai.usage.reasoning_output_tokens", response.usage.output_tokens_details.reasoning_tokens);
                        }
                    }
                    span.setStatus({ code: SpanStatusCode.OK });
                    return response;
                }
                catch (error) {
                    span.setStatus({ code: SpanStatusCode.ERROR, message: error.message });
                    span.recordException(error);
                    throw error;
                }
                finally {
                    span.end();
                }
            }));
        });
    }
}
class EmbeddingsWrapper {
    constructor(openai, tracer) {
        this.openai = openai;
        this.tracer = tracer;
    }
    create(params) {
        return __awaiter(this, void 0, void 0, function* () {
            const { externalCustomerId, externalProductId } = getTracingContext();
            const token = getToken();
            if (!token || !externalCustomerId) {
                throw new Error("No token or externalCustomerId: This wrapper should be used inside a callback to paid.trace().");
            }
            return this.tracer.startActiveSpan("trace.openai.embeddings", (span) => __awaiter(this, void 0, void 0, function* () {
                const attributes = {
                    "gen_ai.system": "openai",
                    "gen_ai.operation.name": "embeddings",
                };
                attributes["external_customer_id"] = externalCustomerId;
                attributes["token"] = token;
                if (externalProductId) {
                    attributes["external_agent_id"] = externalProductId;
                }
                span.setAttributes(attributes);
                try {
                    const response = yield this.openai.embeddings.create(params);
                    if (response.usage) {
                        span.setAttributes({
                            "gen_ai.usage.input_tokens": response.usage.prompt_tokens,
                            "gen_ai.response.model": response.model,
                        });
                    }
                    span.setStatus({ code: SpanStatusCode.OK });
                    return response;
                }
                catch (error) {
                    span.setStatus({ code: SpanStatusCode.ERROR, message: error.message });
                    span.recordException(error);
                    throw error;
                }
                finally {
                    span.end();
                }
            }));
        });
    }
}
class ImagesWrapper {
    constructor(openai, tracer) {
        this.openai = openai;
        this.tracer = tracer;
    }
    generate(params) {
        return __awaiter(this, void 0, void 0, function* () {
            const { externalCustomerId, externalProductId } = getTracingContext();
            const token = getToken();
            const model = params.model || "";
            if (!token || !externalCustomerId) {
                throw new Error("No token or externalCustomerId: This wrapper should be used inside a callback to paid.trace().");
            }
            return this.tracer.startActiveSpan("trace.openai.images", (span) => __awaiter(this, void 0, void 0, function* () {
                var _a;
                const attributes = {
                    "gen_ai.request.model": model,
                    "gen_ai.system": "openai",
                    "gen_ai.operation.name": "image_generation",
                };
                attributes["external_customer_id"] = externalCustomerId;
                attributes["token"] = token;
                if (externalProductId) {
                    attributes["external_agent_id"] = externalProductId;
                }
                span.setAttributes(attributes);
                try {
                    const response = (yield this.openai.images.generate(params));
                    span.setAttributes({
                        "gen_ai.image.count": (_a = params.n) !== null && _a !== void 0 ? _a : 1,
                        "gen_ai.image.size": params.size || "",
                        "gen_ai.image.quality": params.quality || "",
                    });
                    span.setStatus({ code: SpanStatusCode.OK });
                    return response;
                }
                catch (error) {
                    span.setStatus({ code: SpanStatusCode.ERROR, message: error.message });
                    span.recordException(error);
                    throw error;
                }
                finally {
                    span.end();
                }
            }));
        });
    }
}
