// @paid-ai/paid-node/dist/cjs/Client.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaidClient = void 0;
const core = __importStar(require("./core/index.js"));
const headers_js_1 = require("./core/headers.js");
const Client_js_1 = require("./api/resources/customers/client/Client.js");
const Client_js_2 = require("./api/resources/agents/client/Client.js");
const Client_js_3 = require("./api/resources/products/client/Client.js");
const Client_js_4 = require("./api/resources/contacts/client/Client.js");
const Client_js_5 = require("./api/resources/orders/client/Client.js");
const Client_js_6 = require("./api/resources/plans/client/Client.js");
const BatchUsage_js_1 = require("./wrapper/BatchUsage.js");
const Client_js_7 = require("./api/resources/traces/client/Client.js");
const signal_js_1 = require("./tracing/signal.js");
const tracing_js_1 = require("./tracing/tracing.js");
class PaidClient {
    constructor(_options = {}) {
        this._options = Object.assign(Object.assign({}, _options), { headers: (0, headers_js_1.mergeHeaders)({
                "X-Fern-Language": "JavaScript",
                "X-Fern-SDK-Name": "",
                "X-Fern-SDK-Version": "0.8.0",
                "X-Fern-Runtime": core.RUNTIME.type,
                "X-Fern-Runtime-Version": core.RUNTIME.version,
            }, _options === null || _options === void 0 ? void 0 : _options.headers) });
    }
    get customers() {
        var _a;
        return ((_a = this._customers) !== null && _a !== void 0 ? _a : (this._customers = new Client_js_1.Customers(this._options)));
    }
    get agents() {
        var _a;
        return ((_a = this._agents) !== null && _a !== void 0 ? _a : (this._agents = new Client_js_2.Agents(this._options)));
    }
    get products() {
        var _a;
        return ((_a = this._products) !== null && _a !== void 0 ? _a : (this._products = new Client_js_3.Products(this._options)));
    }
    get contacts() {
        var _a;
        return ((_a = this._contacts) !== null && _a !== void 0 ? _a : (this._contacts = new Client_js_4.Contacts(this._options)));
    }
    get orders() {
        var _a;
        return ((_a = this._orders) !== null && _a !== void 0 ? _a : (this._orders = new Client_js_5.Orders(this._options)));
    }
    get plans() {
        var _a;
        return ((_a = this._plans) !== null && _a !== void 0 ? _a : (this._plans = new Client_js_6.Plans(this._options)));
    }
    get usage() {
        var _a;
        return ((_a = this._usage) !== null && _a !== void 0 ? _a : (this._usage = new BatchUsage_js_1.Usage(this._options)));
    }
    get traces() {
        var _a;
        return ((_a = this._traces) !== null && _a !== void 0 ? _a : (this._traces = new Client_js_7.Traces(this._options)));
    }
    /**
     * @deprecated use the standalone 'initializeTracing' instead
     */
    initializeTracing(collectorEndpoint) {
        return __awaiter(this, void 0, void 0, function* () {
            const tokenSupplier = this._options.token;
            const token = typeof tokenSupplier === "function" ? yield tokenSupplier() : tokenSupplier;
            const resolvedToken = yield Promise.resolve(token);
            (0, tracing_js_1.initializeTracing)(resolvedToken, collectorEndpoint);
        });
    }
    /**
     * @deprecated use the standalone 'trace()' instead
     * Use this method to track actions like LLM costs and sending signals.
     * The callback to this function is the work that you want to trace.
     *
     * @param externalCustomerId - The external customer ID
     * @param fn - The callback function containing the work to be traced
     * @param externalAgentId - @deprecated Use externalProductId instead
     * @param externalProductId - The external product ID
     * @param storePrompt - Whether to store the prompt (defaults to false)
     * @param args - Arguments to pass to the callback function
     * @returns The return value of the callback function
     */
    trace(externalCustomerId_1, fn_1, externalAgentId_1, externalProductId_1) {
        return __awaiter(this, arguments, void 0, function* (externalCustomerId, fn, externalAgentId, externalProductId, storePrompt = false, ...args) {
            // TODO: remove this after deprecation and replace externalAgentId with externalProductId everywhere
            if (externalProductId) {
                externalAgentId = externalProductId;
            }
            return yield (0, tracing_js_1.trace)({ externalCustomerId, externalProductId: externalAgentId, storePrompt }, fn, ...args);
        });
    }
    signal(eventName, enableCostTracingOrData, data) {
        let enableCostTracing = false;
        let finalData;
        if (typeof enableCostTracingOrData === "boolean") {
            // Case: signal(eventName, boolean, data?)
            enableCostTracing = enableCostTracingOrData;
            finalData = data;
        }
        else if (typeof enableCostTracingOrData === "object") {
            // Case: signal(eventName, data)
            enableCostTracing = false;
            finalData = enableCostTracingOrData;
        }
        // Case: signal(eventName) - both remain default/undefined
        return (0, signal_js_1.signal)(eventName, enableCostTracing, finalData);
    }
    /**
     * @deprecated use the standalone 'getPaidTracerProvider()' instead
     * Export the tracer provider which user can use for his own tracing.
     */
    get tracerProvider() {
        return (0, tracing_js_1.getPaidTracerProvider)();
    }
}
exports.PaidClient = PaidClient;
