// @paid-ai/paid-node/dist/cjs/integrations/adapters/expressjs/routes.js
"use strict";
/**
 * ExpressJS route helpers for Paid SDK
 * These functions create handlers compatible with ExpressJS controllers
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.createActivateOrderSyncRoute = createActivateOrderSyncRoute;
exports.createContactsRoute = createContactsRoute;
exports.createCustomerInvoicesRoute = createCustomerInvoicesRoute;
exports.createGetCustomerRoute = createGetCustomerRoute;
exports.createCustomersRoute = createCustomersRoute;
exports.createOrdersRoute = createOrdersRoute;
exports.createPayInvoiceRoute = createPayInvoiceRoute;
exports.createProvisioningRoute = createProvisioningRoute;
exports.createSetupIntentRoute = createSetupIntentRoute;
const base_adapter_js_1 = require("./base-adapter.js");
const contacts_js_1 = require("../../controllers/contacts.js");
const invoices_js_1 = require("../../controllers/invoices.js");
const customers_js_1 = require("../../controllers/customers.js");
const orders_js_1 = require("../../controllers/orders.js");
const provision_users_js_1 = require("../../controllers/provision-users.js");
const billing_js_1 = require("../../controllers/billing.js");
/**
 * Helper to extract headers from Express Request
 */
function extractHeaders(req) {
    const headers = {};
    Object.entries(req.headers).forEach(([key, value]) => {
        if (typeof value === "string") {
            headers[key] = value;
        }
        else if (Array.isArray(value)) {
            headers[key] = value.join(", ");
        }
    });
    return headers;
}
/**
 * Create an Express handler for synchronous order activation
 *
 * Use this in your Express app to handle order activation with payment.
 *
 * @example
 * ```typescript
 * import express from 'express';
 * import { createActivateOrderSyncRoute } from '@paid-ai/paid-node/integrations/express';
 *
 * const app = express();
 * const activateOrderSyncHandler = createActivateOrderSyncRoute();
 *
 * app.post('/paid/payments/activate-order-sync', activateOrderSyncHandler);
 * ```
 *
 * @example With NestJS
 * ```typescript
 * import { Controller, Post, Req, Res } from '@nestjs/common';
 * import { Request, Response } from 'express';
 * import { createActivateOrderSyncRoute } from '@paid-ai/paid-node/integrations/nestjs';
 *
 * @Controller('paid/payments')
 * export class PaymentsController {
 *   private activateOrderSyncHandler = createActivateOrderSyncRoute();
 *
 *   @Post('activate-order-sync')
 *   async activateOrderSync(@Req() req: Request, @Res() res: Response) {
 *     return this.activateOrderSyncHandler(req, res);
 *   }
 * }
 * ```
 */
function createActivateOrderSyncRoute(config = {}) {
    const handler = (0, billing_js_1.createActivateOrderSyncHandler)(config.defaultReturnUrl);
    return (req, res) => {
        return handler({
            body: req.body,
            headers: extractHeaders(req),
            method: "POST",
            params: req.params,
        }, (0, base_adapter_js_1.createExpressJSResponseContext)(res), config);
    };
}
/**
 * Create an Express handler for contact creation
 *
 * @example
 * ```typescript
 * import express from 'express';
 * import { createContactsRoute } from '@paid-ai/paid-node/integrations/express';
 *
 * const app = express();
 * app.post('/paid/contacts', createContactsRoute());
 * ```
 */
function createContactsRoute(config = {}) {
    const handler = (0, contacts_js_1.createContactsHandler)();
    return (req, res) => {
        return handler({
            body: req.body,
            headers: extractHeaders(req),
            method: "POST",
            params: req.params,
        }, (0, base_adapter_js_1.createExpressJSResponseContext)(res), config);
    };
}
/**
 * Create an Express handler for fetching customer invoices
 *
 * @example
 * ```typescript
 * import express from 'express';
 * import { createCustomerInvoicesRoute } from '@paid-ai/paid-node/integrations/express';
 *
 * const app = express();
 * app.get('/paid/customers/:customerExternalId/invoices', createCustomerInvoicesRoute());
 * ```
 */
function createCustomerInvoicesRoute(config = {}) {
    const handler = (0, invoices_js_1.createCustomerInvoicesHandler)();
    return (req, res) => {
        return handler({
            body: req.body,
            headers: extractHeaders(req),
            method: "GET",
            params: req.params,
        }, (0, base_adapter_js_1.createExpressJSResponseContext)(res), config);
    };
}
/**
 * Create an Express handler for fetching a customer by external ID
 *
 * @example
 * ```typescript
 * import express from 'express';
 * import { createGetCustomerRoute } from '@paid-ai/paid-node/integrations/express';
 *
 * const app = express();
 * app.get('/paid/customers/:customerExternalId', createGetCustomerRoute());
 * ```
 */
function createGetCustomerRoute(config = {}) {
    const handler = (0, customers_js_1.createGetCustomerHandler)();
    return (req, res) => {
        return handler({
            body: req.body,
            headers: extractHeaders(req),
            method: "GET",
            params: req.params,
        }, (0, base_adapter_js_1.createExpressJSResponseContext)(res), config);
    };
}
/**
 * Create an Express handler for customer creation
 *
 * @example
 * ```typescript
 * import express from 'express';
 * import { createCustomersRoute } from '@paid-ai/paid-node/integrations/express';
 *
 * const app = express();
 * app.post('/paid/customers', createCustomersRoute());
 * ```
 */
function createCustomersRoute(config = {}) {
    const handler = (0, customers_js_1.createCustomersHandler)();
    return (req, res) => {
        return handler({
            body: req.body,
            headers: extractHeaders(req),
            method: "POST",
            params: req.params,
        }, (0, base_adapter_js_1.createExpressJSResponseContext)(res), config);
    };
}
/**
 * Create an Express handler for order creation
 *
 * @example
 * ```typescript
 * import express from 'express';
 * import { createOrdersRoute } from '@paid-ai/paid-node/integrations/express';
 *
 * const app = express();
 * app.post('/paid/orders', createOrdersRoute({ helperOptions: { autoActivate: true } }));
 * ```
 */
function createOrdersRoute(config = {}) {
    const handler = (0, orders_js_1.createOrdersHandler)(config.helperOptions);
    return (req, res) => {
        return handler({
            body: req.body,
            headers: extractHeaders(req),
            method: "POST",
            params: req.params,
        }, (0, base_adapter_js_1.createExpressJSResponseContext)(res), config);
    };
}
/**
 * Create an Express handler for invoice payment
 *
 * @example
 * ```typescript
 * import express from 'express';
 * import { createPayInvoiceRoute } from '@paid-ai/paid-node/integrations/express';
 *
 * const app = express();
 * app.post('/paid/invoices/:invoiceId/pay', createPayInvoiceRoute());
 * ```
 */
function createPayInvoiceRoute(config = {}) {
    const handler = (0, billing_js_1.createPayInvoiceHandler)();
    return (req, res) => {
        return handler({
            body: req.body,
            headers: extractHeaders(req),
            method: "POST",
            params: req.params,
        }, (0, base_adapter_js_1.createExpressJSResponseContext)(res), config);
    };
}
/**
 * Create an Express handler for user provisioning
 *
 * @example
 * ```typescript
 * import express from 'express';
 * import { createProvisioningRoute } from '@paid-ai/paid-node/integrations/express';
 *
 * const app = express();
 * app.post('/paid/provision', createProvisioningRoute({
 *   defaultAgentExternalId: process.env.PAID_AGENT_ID
 * }));
 * ```
 */
function createProvisioningRoute(config = {}) {
    const handler = (0, provision_users_js_1.createProvisioningHandler)(config.orderOptions, config.defaultAgentExternalId);
    return (req, res) => {
        return handler({
            body: req.body,
            headers: extractHeaders(req),
            method: "POST",
            params: req.params,
        }, (0, base_adapter_js_1.createExpressJSResponseContext)(res), config);
    };
}
/**
 * Create an Express handler for setup intent creation
 *
 * @example
 * ```typescript
 * import express from 'express';
 * import { createSetupIntentRoute } from '@paid-ai/paid-node/integrations/express';
 *
 * const app = express();
 * app.post('/paid/payments/setup-intent', createSetupIntentRoute({
 *   defaultReturnUrl: process.env.APP_URL + '/billing'
 * }));
 * ```
 */
function createSetupIntentRoute(config = {}) {
    const handler = (0, billing_js_1.createSetupIntentHandler)(config.defaultReturnUrl);
    return (req, res) => {
        return handler({
            body: req.body,
            headers: extractHeaders(req),
            method: "POST",
            params: req.params,
        }, (0, base_adapter_js_1.createExpressJSResponseContext)(res), config);
    };
}
