// @paid-ai/paid-node/dist/cjs/integrations/client.js
"use strict";
/**
 * Shared Paid client initialization utility
 * Framework-agnostic
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initializePaidClient = initializePaidClient;
exports.getOrganizationId = getOrganizationId;
const Client_js_1 = require("../Client.js");
/**
 * Initialize a Paid client with environment variable fallbacks
 *
 * @param config - Optional configuration overrides
 * @returns Configured PaidClient instance
 * @throws Error if API key is not provided or found in environment
 *
 * @example
 * ```typescript
 * const client = initializePaidClient();
 * // Uses process.env.PAID_API_KEY
 *
 * const client = initializePaidClient({ apiKey: 'pk_xxx' });
 * // Uses provided key
 * ```
 */
function initializePaidClient(config = {}) {
    const apiKey = config.apiKey || process.env.PAID_API_KEY;
    if (!apiKey) {
        throw new Error('PAID_API_KEY not configured. Set process.env.PAID_API_KEY or pass apiKey in config.');
    }
    return new Client_js_1.PaidClient({
        token: apiKey,
        baseUrl: config.apiUrl
            ? `${config.apiUrl}/api/v1`
            : process.env.PAID_API_URL
                ? `${process.env.PAID_API_URL}/api/v1`
                : undefined,
    });
}
/**
 * Get organization ID from API key
 *
 * Calls the Paid API to resolve the organization ID associated with the API key.
 *
 * @param config - API configuration
 * @returns Organization ID or null if not found
 *
 * @example
 * ```typescript
 * const orgId = await getOrganizationId({
 *   apiKey: process.env.PAID_API_KEY,
 *   apiUrl: 'https://api.agentpaid.io'
 * });
 * ```
 */
function getOrganizationId() {
    return __awaiter(this, arguments, void 0, function* (config = {}) {
        var _a;
        const apiKey = config.apiKey || process.env.PAID_API_KEY;
        const apiUrl = config.apiUrl || process.env.PAID_API_URL || 'https://api.agentpaid.io';
        if (!apiKey) {
            throw new Error('PAID_API_KEY not configured');
        }
        try {
            const response = yield fetch(`${apiUrl}/api/organizations/organizationId`, {
                method: 'GET',
                headers: {
                    Authorization: `Bearer ${apiKey}`,
                    'Content-Type': 'application/json',
                },
            });
            if (!response.ok) {
                console.error('Failed to get organization ID:', yield response.text());
                return null;
            }
            const data = yield response.json();
            return ((_a = data.data) === null || _a === void 0 ? void 0 : _a.organizationId) || null;
        }
        catch (error) {
            console.error('Error fetching organization ID:', error);
            return null;
        }
    });
}
