// @paid-ai/paid-node/dist/esm/integrations/controllers/billing.mjs
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { paidApiFetch } from "../utils/api-fetch.mjs";
import { createHandler } from "../utils/base-handler.mjs";
/**
 * Pay an invoice with a payment confirmation token
 * This helper processes payment for an existing invoice using a Stripe confirmation token.
 *
 * @param config - API configuration with organization context
 * @param request - Payment request details
 * @returns Payment result
 *
 * @example
 * ```typescript
 * const result = await payInvoice(
 *   {
 *     apiUrl: 'https://api.agentpaid.io',
 *     apiKey: process.env.PAID_API_KEY,
 *     organizationId: 'org_123'
 *   },
 *   {
 *     invoiceId: 'inv_123',
 *     confirmationToken: 'pm_tok_xxx',
 *     returnUrl: 'https://example.com/thanks'
 *   }
 * );
 * ```
 */
export function payInvoice(config, request) {
    return __awaiter(this, void 0, void 0, function* () {
        const { invoiceId, confirmationToken, returnUrl } = request;
        if (!invoiceId || !confirmationToken) {
            throw new Error("invoiceId and confirmationToken are required");
        }
        const data = yield paidApiFetch(config, `/api/organizations/${config.organizationId}/invoices/${invoiceId}/pay`, {
            method: "PUT",
            body: Object.assign({ confirmationToken }, (returnUrl && { returnUrl })),
        });
        return { success: true, data };
    });
}
/**
 * Create a framework-agnostic handler for invoice payment
 *
 * This handler can be used with any framework adapter.
 * Organization ID is automatically fetched from the API key.
 *
 * @returns Handler function
 */
export function createPayInvoiceHandler() {
    return createHandler((_client, body, params, organizationId) => __awaiter(this, void 0, void 0, function* () {
        if (!organizationId) {
            return { success: false, error: 'Organization ID not found', status: 500 };
        }
        const invoiceId = (params === null || params === void 0 ? void 0 : params.invoiceId) || body.invoiceId;
        const { confirmationToken, returnUrl } = body;
        if (!invoiceId) {
            return { success: false, error: 'invoiceId is required', status: 400 };
        }
        if (!confirmationToken) {
            return { success: false, error: 'confirmationToken is required', status: 400 };
        }
        const apiKey = process.env.PAID_API_KEY || '';
        const apiUrl = process.env.PAID_API_URL || 'https://api.agentpaid.io';
        const result = yield payInvoice({ apiUrl, apiKey, organizationId }, { invoiceId, confirmationToken, returnUrl });
        return {
            success: true,
            data: result.data,
        };
    }), { requiredFields: ['confirmationToken'], requireOrganizationId: true });
}
/**
 * Activate an order synchronously with payment
 *
 * This operation activates an order and immediately charges the customer
 * using the provided confirmation token (from Stripe or other payment provider).
 *
 * @param config - API configuration with organization context
 * @param request - Activation request with order ID and confirmation token
 * @returns Activation result with order and billing information
 *
 * @example
 * ```typescript
 * const result = await activateOrderSync(
 *   {
 *     apiUrl: 'https://api.agentpaid.io',
 *     apiKey: process.env.PAID_API_KEY,
 *     organizationId: 'org_123'
 *   },
 *   {
 *     orderId: 'ord_123',
 *     confirmationToken: 'pi_123_secret_456',
 *     returnUrl: 'https://example.com/return'
 *   }
 * );
 * ```
 */
export function activateOrderSync(config, request) {
    return __awaiter(this, void 0, void 0, function* () {
        const { orderId, confirmationToken, returnUrl } = request;
        if (!orderId) {
            throw new Error("orderId is required");
        }
        if (!confirmationToken) {
            throw new Error("confirmationToken is required");
        }
        const data = yield paidApiFetch(config, `/api/organizations/${config.organizationId}/orders/${orderId}/activate-and-pay`, {
            method: "POST",
            body: Object.assign({ confirmationToken }, (returnUrl && { returnUrl })),
        });
        return {
            success: true,
            order: data.order,
            billing: data.billing,
        };
    });
}
/**
 * Create a framework-agnostic handler for synchronous order activation
 *
 * This handler can be used with any framework adapter.
 *
 * @param defaultReturnUrl - Optional default return URL
 * @returns Handler function
 */
export function createActivateOrderSyncHandler(defaultReturnUrl) {
    return createHandler((client, body, params, organizationId) => __awaiter(this, void 0, void 0, function* () {
        var _a, _b;
        const apiKey = (_a = client._options) === null || _a === void 0 ? void 0 : _a.token;
        const baseUrl = ((_b = client._options) === null || _b === void 0 ? void 0 : _b.baseUrl) || 'https://api.agentpaid.io/api/v1';
        const apiUrl = baseUrl.replace('/api/v1', '');
        const orderId = (params === null || params === void 0 ? void 0 : params.orderId) || body.orderId;
        const result = yield activateOrderSync({ apiUrl, apiKey, organizationId: organizationId }, {
            orderId: orderId,
            confirmationToken: body.confirmationToken,
            returnUrl: body.returnUrl || defaultReturnUrl,
        });
        return {
            success: true,
            data: {
                success: true,
                order: result.order,
                billing: result.billing,
                message: 'Order activated and charged successfully',
            },
        };
    }), {
        requiredFields: ['confirmationToken'],
        requireOrganizationId: true,
    });
}
/**
 * Create a setup intent for adding payment methods without immediate charge
 *
 * Setup intents are used to save payment methods for future use without
 * charging the customer immediately. This is useful for:
 * - Adding payment methods for subscription billing
 * - Updating payment methods
 * - Pre-authorizing payment methods
 *
 * @param config - API configuration with organization context
 * @param request - Setup intent request with customer ID and confirmation token
 * @returns Setup intent result with payment method details
 *
 * @example
 * ```typescript
 * const result = await createSetupIntent(
 *   {
 *     apiUrl: 'https://api.agentpaid.io',
 *     apiKey: process.env.PAID_API_KEY,
 *     organizationId: 'org_123'
 *   },
 *   {
 *     customerId: 'cus_123',
 *     confirmationToken: 'seti_123_secret_456',
 *     returnUrl: 'https://example.com/return',
 *     metadata: { source: 'web' }
 *   }
 * );
 * ```
 */
export function createSetupIntent(config, request) {
    return __awaiter(this, void 0, void 0, function* () {
        const { customerId, confirmationToken, returnUrl, metadata } = request;
        if (!customerId) {
            throw new Error("customerId is required");
        }
        if (!confirmationToken) {
            throw new Error("confirmationToken is required");
        }
        const data = yield paidApiFetch(config, `/api/organizations/${config.organizationId}/payments/setup-intents-external`, {
            method: "POST",
            body: Object.assign(Object.assign({ customerId,
                confirmationToken }, (returnUrl && { returnUrl })), (metadata && { metadata })),
        });
        return {
            setupIntent: data.setupIntent || data,
            message: "Setup intent created successfully",
        };
    });
}
/**
 * Create a framework-agnostic handler for setup intent creation
 *
 * This handler can be used with any framework adapter.
 *
 * @param defaultReturnUrl - Optional default return URL
 * @returns Handler function
 *
 * @example
 * ```typescript
 * // Next.js API route
 * import { createSetupIntentHandler } from '@paid-ai/paid-node/integrations';
 *
 * const handler = createSetupIntentHandler('https://myapp.com/billing');
 *
 * export async function POST(req: Request) {
 *   return handler(
 *     {
 *       body: await req.json(),
 *       headers: Object.fromEntries(req.headers),
 *       method: 'POST'
 *     },
 *     {
 *       json: (data, status = 200) => Response.json(data, { status }),
 *       error: (message, status) => Response.json({ error: message }, { status })
 *     }
 *   );
 * }
 * ```
 */
export function createSetupIntentHandler(defaultReturnUrl) {
    return createHandler((client, body, _params, organizationId) => __awaiter(this, void 0, void 0, function* () {
        var _a, _b;
        const apiKey = (_a = client._options) === null || _a === void 0 ? void 0 : _a.token;
        const baseUrl = ((_b = client._options) === null || _b === void 0 ? void 0 : _b.baseUrl) || 'https://api.agentpaid.io/api/v1';
        const apiUrl = baseUrl.replace('/api/v1', '');
        const result = yield createSetupIntent({ apiUrl, apiKey, organizationId: organizationId }, {
            customerId: body.customerId,
            confirmationToken: body.confirmationToken,
            returnUrl: body.returnUrl || defaultReturnUrl,
            metadata: body.metadata,
        });
        return {
            success: true,
            data: result,
        };
    }), {
        requiredFields: ['customerId', 'confirmationToken'],
        requireOrganizationId: true,
    });
}
