// @paid-ai/paid-node/dist/esm/integrations/controllers/contacts.mjs
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { createHandler } from "../utils/base-handler.mjs";
function generateContactDefaults(data) {
    var _a, _b, _c, _d, _e, _f;
    const timestamp = Date.now();
    const email = data.email || `contact_${timestamp}@example.com`;
    const firstName = data.firstName || "Joe";
    const lastName = data.lastName || "Bloggs";
    const phone = data.phone || "";
    const salutation = data.salutation || "Mr";
    const billingAddress = {
        line1: ((_a = data.billingAddress) === null || _a === void 0 ? void 0 : _a.line1) || "123 Placeholder Street",
        line2: ((_b = data.billingAddress) === null || _b === void 0 ? void 0 : _b.line2) || "",
        city: ((_c = data.billingAddress) === null || _c === void 0 ? void 0 : _c.city) || "London",
        state: ((_d = data.billingAddress) === null || _d === void 0 ? void 0 : _d.state) || "London",
        zipCode: ((_e = data.billingAddress) === null || _e === void 0 ? void 0 : _e.zipCode) || "SW1A 1AA",
        country: ((_f = data.billingAddress) === null || _f === void 0 ? void 0 : _f.country) || "UK",
    };
    return {
        customerExternalId: data.customerExternalId,
        email,
        firstName,
        lastName,
        phone,
        salutation,
        billingAddress,
    };
}
/**
 * Internal helper to create a contact
 */
function createContactInternal(client, data) {
    return __awaiter(this, void 0, void 0, function* () {
        const contact = yield client.contacts.create({
            customerExternalId: data.customerExternalId,
            salutation: data.salutation,
            firstName: data.firstName,
            lastName: data.lastName,
            email: data.email,
            phone: data.phone,
            billingStreet: data.billingAddress.line1,
            billingCity: data.billingAddress.city,
            billingCountry: data.billingAddress.country,
            billingPostalCode: data.billingAddress.zipCode,
        });
        if (!contact.id) {
            throw new Error("Contact created but missing ID");
        }
        return {
            id: contact.id,
        };
    });
}
/**
 * Create a contact with defaults for missing fields
 *
 * This helper generates reasonable placeholder values for any missing data,
 * making it easy to create contacts for prototyping and development.
 *
 * @param client - PaidClient instance
 * @param data - Contact data (only customerExternalId is required)
 * @returns Created contact with ID
 *
 * @example
 * ```typescript
 * const contact = await createContactWithDefaults(client, {
 *   customerExternalId: 'user-123',
 *   email: 'user@example.com'
 * });
 * // Generates: firstName, lastName, address, etc.
 * ```
 */
export function createContactWithDefaults(client, data) {
    return __awaiter(this, void 0, void 0, function* () {
        if (!data.customerExternalId) {
            throw new Error("customerExternalId is required");
        }
        const completeData = generateContactDefaults(data);
        return createContactInternal(client, completeData);
    });
}
/**
 * Create a contact with explicit data (NO defaults applied)
 *
 * Required fields: customerExternalId, email, firstName, lastName, phone, salutation,
 * and complete billingAddress (line1, line2, city, state, zipCode, country)
 *
 * Use createContactWithDefaults() if you want automatic default values for missing fields.
 *
 * @param client - PaidClient instance
 * @param data - Complete contact data with ALL required fields
 * @returns Created contact with ID
 * @throws Error if any required fields are missing
 *
 * @example
 * ```typescript
 * const contact = await createContact(client, {
 *   customerExternalId: 'user-123',
 *   email: 'john@example.com',
 *   firstName: 'John',
 *   lastName: 'Doe',
 *   phone: '+1234567890',
 *   salutation: 'Mr',
 *   billingAddress: {
 *     line1: '123 Main St',
 *     line2: 'Suite 100',
 *     city: 'San Francisco',
 *     state: 'CA',
 *     zipCode: '94105',
 *     country: 'US'
 *   }
 * });
 * ```
 */
export function createContact(client, data) {
    return __awaiter(this, void 0, void 0, function* () {
        return createContactInternal(client, data);
    });
}
/**
 * Create a framework-agnostic handler for contact creation
 *
 * This handler can be used with any framework adapter.
 *
 * @returns Handler function
 */
export function createContactsHandler() {
    return createHandler((client, body) => __awaiter(this, void 0, void 0, function* () {
        const contact = yield createContactWithDefaults(client, body);
        return { success: true, data: contact };
    }), { requiredFields: ['customerExternalId'] });
}
