// @paid-ai/paid-node/dist/cjs/integrations/controllers/customers.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createCustomerWithDefaults = createCustomerWithDefaults;
exports.createCustomer = createCustomer;
exports.createCustomersHandler = createCustomersHandler;
exports.getCustomer = getCustomer;
exports.createGetCustomerHandler = createGetCustomerHandler;
const base_handler_js_1 = require("../utils/base-handler.js");
/**
 * Generate placeholder values for missing customer fields
 * Creates valid but fake data for rapid prototyping
 */
function generateCustomerDefaults(data) {
    var _a, _b, _c, _d, _e, _f;
    const timestamp = Date.now();
    const randomId = Math.random().toString(36).substring(7);
    const email = data.email || `user_${timestamp}@example.com`;
    const name = data.name || `User ${randomId}`;
    const nameParts = name.split(" ");
    const firstName = data.firstName || nameParts[0] || "Joe";
    const lastName = data.lastName || nameParts.slice(1).join(" ") || "Bloggs";
    const billingAddress = {
        line1: ((_a = data.billingAddress) === null || _a === void 0 ? void 0 : _a.line1) || "123 Placeholder Street",
        line2: ((_b = data.billingAddress) === null || _b === void 0 ? void 0 : _b.line2) || "",
        city: ((_c = data.billingAddress) === null || _c === void 0 ? void 0 : _c.city) || "London",
        state: ((_d = data.billingAddress) === null || _d === void 0 ? void 0 : _d.state) || "London",
        zipCode: ((_e = data.billingAddress) === null || _e === void 0 ? void 0 : _e.zipCode) || "SW1A 1AA",
        country: ((_f = data.billingAddress) === null || _f === void 0 ? void 0 : _f.country) || "UK",
    };
    return {
        externalId: data.externalId,
        email,
        name,
        firstName,
        lastName,
        company: data.company || "",
        phone: data.phone || "",
        billingAddress,
        metadata: data.metadata || {},
    };
}
/**
 * Internal helper to create a customer
 */
function createCustomerInternal(client, data) {
    return __awaiter(this, void 0, void 0, function* () {
        const customer = yield client.customers.create({
            externalId: data.externalId,
            name: data.name,
            billingAddress: {
                line1: data.billingAddress.line1,
                line2: data.billingAddress.line2,
                city: data.billingAddress.city,
                state: data.billingAddress.state,
                zipCode: data.billingAddress.zipCode,
                country: data.billingAddress.country,
            },
        });
        if (!customer.id) {
            throw new Error("Customer created but missing ID");
        }
        return {
            id: customer.id,
            externalId: data.externalId,
        };
    });
}
/**
 * Create a customer with defaults for missing fields
 *
 * This helper generates reasonable placeholder values for any missing data,
 * making it easy to create customers for prototyping and development.
 *
 * @param client - PaidClient instance
 * @param data - Customer data (only externalId is required)
 * @returns Created customer with ID
 *
 * @example
 * ```typescript
 * const customer = await createCustomerWithDefaults(client, {
 *   externalId: 'user-123',
 *   email: 'user@example.com'
 * });
 * // Generates: name, address, etc.
 * ```
 */
function createCustomerWithDefaults(client, data) {
    return __awaiter(this, void 0, void 0, function* () {
        const completeData = generateCustomerDefaults(data);
        return createCustomerInternal(client, completeData);
    });
}
/**
 * Create a customer with explicit data (NO defaults applied)
 *
 * Required fields: externalId, name, email, and complete billingAddress
 * (line1, line2, city, state, zipCode, country)
 *
 * Use createCustomerWithDefaults() if you want automatic default values for missing fields.
 *
 * @param client - PaidClient instance
 * @param data - Complete customer data with ALL required fields
 * @returns Created customer with ID
 * @throws Error if any required fields are missing
 *
 * @example
 * ```typescript
 * const customer = await createCustomer(client, {
 *   externalId: 'user-123',
 *   email: 'user@example.com',
 *   name: 'John Doe',
 *   billingAddress: {
 *     line1: '123 Main St',
 *     line2: '',
 *     city: 'San Francisco',
 *     state: 'CA',
 *     zipCode: '94105',
 *     country: 'US'
 *   }
 * });
 * ```
 */
function createCustomer(client, data) {
    return __awaiter(this, void 0, void 0, function* () {
        return createCustomerInternal(client, data);
    });
}
/**
 * Create a framework-agnostic handler for customer creation
 *
 * This handler can be used with any framework adapter.
 *
 * @returns Handler function
 */
function createCustomersHandler() {
    return (0, base_handler_js_1.createHandler)((client, body) => __awaiter(this, void 0, void 0, function* () {
        if (!body.externalId) {
            throw new Error("externalId is required");
        }
        const customer = yield createCustomerWithDefaults(client, body);
        return { success: true, data: customer };
    }), { requiredFields: ['externalId'] });
}
/**
 * Get a customer by external ID
 *
 * @param client - PaidClient instance
 * @param customerExternalId - Customer's external ID
 * @returns Customer data
 *
 * @example
 * ```typescript
 * const customer = await getCustomer(client, 'user-123');
 * console.log(customer.name, customer.email);
 * ```
 */
function getCustomer(client, customerExternalId) {
    return __awaiter(this, void 0, void 0, function* () {
        const customer = yield client.customers.getByExternalId(customerExternalId);
        return customer;
    });
}
/**
 * Create a framework-agnostic handler for fetching a customer by external ID
 *
 * This handler can be used with any framework adapter.
 * The customer external ID should be provided either in params or query string.
 *
 * @returns Handler function
 *
 * @example
 * ```typescript
 * // In Next.js route: /api/customers/[customerExternalId]/route.ts
 * import { createGetCustomerHandler } from '@paid-ai/paid-node/integrations';
 * import { nextjsAdapter } from '@paid-ai/paid-node/integrations/nextjs';
 *
 * const handler = createGetCustomerHandler();
 * export const GET = nextjsAdapter(handler);
 * ```
 */
function createGetCustomerHandler() {
    return (0, base_handler_js_1.createHandler)((client, _body, params) => __awaiter(this, void 0, void 0, function* () {
        const customerExternalId = params === null || params === void 0 ? void 0 : params.customerExternalId;
        if (!customerExternalId) {
            return {
                success: false,
                error: "customerExternalId is required",
                status: 400,
            };
        }
        const customer = yield getCustomer(client, customerExternalId);
        return { success: true, data: customer };
    }), {
        allowedMethods: ['GET'],
        requireOrganizationId: false,
    });
}
