// @paid-ai/paid-node/dist/cjs/integrations/controllers/orders.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createOrderWithDefaults = createOrderWithDefaults;
exports.createOrder = createOrder;
exports.createOrdersHandler = createOrdersHandler;
const base_handler_js_1 = require("../utils/base-handler.js");
/**
 * Generate default values for missing order fields
 */
function generateOrderDefaults(config, options = {}) {
    var _a;
    const { defaultDurationDays = 365 } = options;
    const startDate = config.startDate || new Date().toISOString().split("T")[0];
    const endDate = config.endDate ||
        new Date(Date.now() + defaultDurationDays * 24 * 60 * 60 * 1000)
            .toISOString()
            .split("T")[0];
    const name = config.name || "Subscription Order";
    const description = config.description || "Annual subscription";
    const currency = config.currency || "USD";
    const orderLines = ((_a = config.orderLines) === null || _a === void 0 ? void 0 : _a.map(line => ({
        agentExternalId: line.agentExternalId,
        name: line.name || name,
        description: line.description || description,
    }))) || [{
            agentExternalId: config.agentExternalId,
            name,
            description,
        }];
    return {
        customerId: config.customerId,
        customerExternalId: config.customerExternalId,
        billingContactId: config.billingContactId,
        agentExternalId: config.agentExternalId,
        name,
        description,
        startDate,
        endDate,
        currency,
        orderLines,
    };
}
/**
 * Internal helper to create and optionally activate an order
 */
function createAndActivateOrder(client, config, autoActivate) {
    return __awaiter(this, void 0, void 0, function* () {
        const order = yield client.orders.create({
            customerId: config.customerId,
            customerExternalId: config.customerExternalId,
            billingContactId: config.billingContactId,
            name: config.name,
            description: config.description,
            startDate: config.startDate,
            endDate: config.endDate,
            currency: config.currency,
            orderLines: config.orderLines.map(line => {
                const orderLine = { agentExternalId: line.agentExternalId };
                if (line.name)
                    orderLine.name = line.name;
                if (line.description)
                    orderLine.description = line.description;
                return orderLine;
            }),
        });
        if (!order.id) {
            throw new Error("Order created but missing ID");
        }
        if (autoActivate && order.creationState === "draft") {
            try {
                yield client.orders.activate(order.id);
                return {
                    id: order.id,
                    creationState: "active",
                };
            }
            catch (activationError) {
                console.error(`Error activating order ${order.id}:`, activationError instanceof Error
                    ? activationError.message
                    : "Unknown error");
                return {
                    id: order.id,
                    creationState: order.creationState,
                };
            }
        }
        return {
            id: order.id,
            creationState: order.creationState,
        };
    });
}
/**
 * Create an order with defaults for missing fields
 *
 * This helper generates reasonable defaults for dates, names, and other fields,
 * making it easy to create orders for prototyping and development.
 *
 * @param client - PaidClient instance
 * @param config - Order configuration (customerId, customerExternalId, agentExternalId required)
 * @param options - Additional options for order creation
 * @returns Created order with ID
 *
 * @example
 * ```typescript
 * const order = await createOrderWithDefaults(client, {
 *   customerId: 'cus_123',
 *   customerExternalId: 'user-123',
 *   agentExternalId: 'agent-prod'
 * });
 * // Generates: startDate, endDate (1 year), name, description, currency (USD)
 * ```
 */
function createOrderWithDefaults(client_1, config_1) {
    return __awaiter(this, arguments, void 0, function* (client, config, options = {}) {
        const { autoActivate = true } = options;
        const completeConfig = generateOrderDefaults(config, options);
        return createAndActivateOrder(client, completeConfig, autoActivate);
    });
}
/**
 * Create an order with explicit configuration (NO defaults applied)
 *
 * Required fields: customerId, customerExternalId, agentExternalId, name, description,
 * startDate, endDate, currency, orderLines
 *
 * Use createOrderWithDefaults() if you want automatic default values for missing fields.
 *
 * @param client - PaidClient instance
 * @param config - Complete order configuration with ALL required fields
 * @param options - Additional options for order creation
 * @returns Created order with ID
 * @throws Error if any required fields are missing
 *
 * @example
 * ```typescript
 * const order = await createOrder(client, {
 *   customerId: 'cus_123',
 *   customerExternalId: 'user-123',
 *   billingContactId: 'contact_123',
 *   agentExternalId: 'agent-prod',
 *   name: 'Pro Plan Subscription',
 *   description: 'Monthly subscription to Pro plan',
 *   startDate: '2024-01-01',
 *   endDate: '2025-01-01',
 *   currency: 'USD',
 *   orderLines: [{
 *     agentExternalId: 'agent-prod',
 *     name: 'Pro Plan',
 *     description: 'Pro plan features'
 *   }]
 * }, { autoActivate: true });
 * ```
 */
function createOrder(client_1, config_1) {
    return __awaiter(this, arguments, void 0, function* (client, config, options = {}) {
        const { autoActivate = true } = options;
        return createAndActivateOrder(client, config, autoActivate);
    });
}
/**
 * Create a framework-agnostic handler for order creation
 *
 * This handler can be used with any framework adapter.
 *
 * @param helperOptions - Default options for order creation
 * @returns Handler function
 */
function createOrdersHandler(helperOptions) {
    return (0, base_handler_js_1.createHandler)((client, body) => __awaiter(this, void 0, void 0, function* () {
        var _a, _b;
        if (!body.customerId) {
            throw new Error("customerId is required");
        }
        if (!body.customerExternalId) {
            throw new Error("customerExternalId is required");
        }
        if (!body.agentExternalId) {
            throw new Error("agentExternalId is required");
        }
        const orderOptions = Object.assign(Object.assign({}, helperOptions), { autoActivate: (_b = (_a = body.autoActivate) !== null && _a !== void 0 ? _a : helperOptions === null || helperOptions === void 0 ? void 0 : helperOptions.autoActivate) !== null && _b !== void 0 ? _b : true });
        const order = yield createOrderWithDefaults(client, body, orderOptions);
        return { success: true, data: order };
    }), { requiredFields: ['customerId', 'customerExternalId', 'agentExternalId'] });
}
