// @paid-ai/paid-node/dist/cjs/integrations/controllers/provision-users.js
"use strict";
/**
 * Provisioning helpers for Paid SDK
 * Orchestrates customer + contact + order creation in one operation
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.provisionNewUser = provisionNewUser;
exports.createProvisioningHandler = createProvisioningHandler;
const customers_js_1 = require("./customers.js");
const contacts_js_1 = require("./contacts.js");
const orders_js_1 = require("./orders.js");
const base_handler_js_1 = require("../utils/base-handler.js");
/**
 * Provision a new user with customer, contact, and optionally order
 *
 * This is a high-level helper that orchestrates the creation of all
 * necessary Paid entities for a new user in one operation.
 *
 * @param client - PaidClient instance
 * @param config - Provisioning configuration
 * @returns Result with created customer, contact, and optional order
 *
 * @example
 * ```typescript
 * const result = await provisionNewUser(client, {
 *   customer: {
 *     externalId: 'user-123',
 *     email: 'user@example.com'
 *   },
 *   agentExternalId: 'agent-prod'
 * });
 * // Creates customer, contact, and order with defaults
 * ```
 */
function provisionNewUser(client, config) {
    return __awaiter(this, void 0, void 0, function* () {
        var _a, _b, _c, _d, _e, _f;
        const customer = yield (0, customers_js_1.createCustomerWithDefaults)(client, config.customer);
        const contactData = {
            customerExternalId: config.customer.externalId,
            email: ((_a = config.contact) === null || _a === void 0 ? void 0 : _a.email) || config.customer.email,
            firstName: (_b = config.contact) === null || _b === void 0 ? void 0 : _b.firstName,
            lastName: (_c = config.contact) === null || _c === void 0 ? void 0 : _c.lastName,
            phone: (_d = config.contact) === null || _d === void 0 ? void 0 : _d.phone,
            billingAddress: (_e = config.contact) === null || _e === void 0 ? void 0 : _e.billingAddress,
            salutation: (_f = config.contact) === null || _f === void 0 ? void 0 : _f.salutation,
        };
        const contact = yield (0, contacts_js_1.createContactWithDefaults)(client, contactData);
        let order;
        if (config.agentExternalId) {
            order = yield (0, orders_js_1.createOrderWithDefaults)(client, {
                customerId: customer.id,
                customerExternalId: config.customer.externalId,
                agentExternalId: config.agentExternalId,
                billingContactId: contact.id,
            }, config.orderOptions);
        }
        return {
            customer,
            contact,
            order,
        };
    });
}
/**
 * Create a framework-agnostic handler for user provisioning
 *
 * This handler can be used with any framework adapter.
 * Accepts partial provisioning configuration and fills in defaults for missing fields.
 *
 * @param orderOptions - Optional default order creation options
 * @param defaultAgentExternalId - Optional default agent external ID
 * @returns Handler function
 *
 * @example
 * ```typescript
 * // Minimal - just externalId and email
 * {
 *   customer: { externalId: 'org-123', email: 'user@example.com' }
 * }
 *
 * // With full customer data
 * {
 *   customer: {
 *     externalId: 'org-123',
 *     email: 'user@example.com',
 *     name: 'Acme Corp',
 *     billingAddress: { line1: '123 Main St', city: 'NYC', ... }
 *   },
 *   contact: {
 *     firstName: 'John',
 *     lastName: 'Doe'
 *   }
 * }
 * ```
 */
function createProvisioningHandler(orderOptions, defaultAgentExternalId) {
    return (0, base_handler_js_1.createHandler)((client, body) => __awaiter(this, void 0, void 0, function* () {
        if (!body.customer) {
            return {
                success: false,
                error: "'customer' is required",
                status: 400,
            };
        }
        const effectiveAgentId = body.agentExternalId ||
            defaultAgentExternalId ||
            process.env.NEXT_PUBLIC_PAID_AGENT_ID ||
            process.env.PAID_AGENT_ID;
        const result = yield provisionNewUser(client, {
            customer: body.customer,
            contact: body.contact,
            agentExternalId: effectiveAgentId,
            orderOptions: body.orderOptions || orderOptions,
        });
        return {
            success: true,
            data: result,
        };
    }), { requiredFields: [] });
}
