// @paid-ai/paid-node/dist/cjs/tracing/wrappers/anthropicWrapper.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PaidAnthropic = void 0;
const api_1 = require("@opentelemetry/api");
const tracing_js_1 = require("tracing/tracing.js");
const tracingContext_js_1 = require("tracing/tracingContext.js");
class PaidAnthropic {
    constructor(anthropicClient) {
        this.anthropic = anthropicClient;
        const tracer = (0, tracing_js_1.getPaidTracer)();
        if (!tracer) {
            throw new Error("Paid tracer is not initialized, Make sure to call 'initializeTracing()' first");
        }
        this.tracer = tracer;
    }
    get messages() {
        return new MessagesWrapper(this.anthropic, this.tracer);
    }
}
exports.PaidAnthropic = PaidAnthropic;
class MessagesWrapper {
    constructor(anthropic, tracer) {
        this.anthropic = anthropic;
        this.tracer = tracer;
    }
    create(params) {
        return __awaiter(this, void 0, void 0, function* () {
            const token = (0, tracing_js_1.getToken)();
            const { externalProductId, externalCustomerId } = (0, tracingContext_js_1.getTracingContext)();
            if (!token || !externalCustomerId) {
                throw new Error("No token or externalCustomerId: This wrapper should be used inside a callback to paid.trace().");
            }
            return this.tracer.startActiveSpan("trace.anthropic.messages", (span) => __awaiter(this, void 0, void 0, function* () {
                const attributes = {
                    "gen_ai.system": "anthropic",
                    "gen_ai.operation.name": "messages",
                    external_customer_id: externalCustomerId,
                    token: token,
                };
                if (externalProductId) {
                    attributes["external_agent_id"] = externalProductId;
                }
                // Set request model from params
                if (params.model) {
                    attributes["gen_ai.request.model"] = params.model;
                }
                span.setAttributes(attributes);
                try {
                    const response = (yield this.anthropic.messages.create(params));
                    // Add usage information from response
                    if (response.usage) {
                        span.setAttributes({
                            "gen_ai.usage.input_tokens": response.usage.input_tokens,
                            "gen_ai.usage.output_tokens": response.usage.output_tokens,
                            "gen_ai.response.model": response.model,
                        });
                        // Add Anthropic-specific cache usage if available
                        if ("cache_creation_input_tokens" in response.usage && response.usage.cache_creation_input_tokens) {
                            span.setAttribute("gen_ai.usage.cache_creation_input_tokens", response.usage.cache_creation_input_tokens);
                        }
                        if ("cache_read_input_tokens" in response.usage && response.usage.cache_read_input_tokens) {
                            span.setAttribute("gen_ai.usage.cache_read_input_tokens", response.usage.cache_read_input_tokens);
                        }
                    }
                    span.setStatus({ code: api_1.SpanStatusCode.OK });
                    return response;
                }
                catch (error) {
                    span.setStatus({ code: api_1.SpanStatusCode.ERROR, message: error.message });
                    span.recordException(error);
                    throw error;
                }
                finally {
                    span.end();
                }
            }));
        });
    }
}
