// @paid-ai/paid-node/dist/cjs/integrations/adapters/nextjs/routes.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createActivateOrderSyncRoute = createActivateOrderSyncRoute;
exports.createContactsRoute = createContactsRoute;
exports.createCustomerInvoicesRoute = createCustomerInvoicesRoute;
exports.createGetCustomerRoute = createGetCustomerRoute;
exports.createCustomersRoute = createCustomersRoute;
exports.createOrdersRoute = createOrdersRoute;
exports.createPayInvoiceRoute = createPayInvoiceRoute;
exports.createProvisioningRoute = createProvisioningRoute;
exports.createSetupIntentRoute = createSetupIntentRoute;
const base_adapter_js_1 = require("./base-adapter.js");
const contacts_js_1 = require("../../controllers/contacts.js");
const invoices_js_1 = require("../../controllers/invoices.js");
const customers_js_1 = require("../../controllers/customers.js");
const orders_js_1 = require("../../controllers/orders.js");
const provision_users_js_1 = require("../../controllers/provision-users.js");
const billing_js_1 = require("../../controllers/billing.js");
/**
 * Create a Next.js API route handler for synchronous order activation
 *
 * This creates an endpoint that activates an order and immediately charges
 * the customer using a confirmation token from a payment provider.
 *
 * Organization ID is automatically fetched from the API key.
 *
 * @example
 * ```typescript
 * // src/app/api/orders/[orderId]/activate-sync/route.ts
 * import { createActivateOrderSyncRoute } from '@paid-ai/paid-node/integrations/nextjs';
 *
 * export const POST = createActivateOrderSyncRoute({
 *   defaultReturnUrl: process.env.NEXT_PUBLIC_BASE_URL + '/orders'
 * });
 * ```
 *
 * @example Client-side usage with route params
 * ```typescript
 * await fetch('/api/orders/ord_123/activate-sync', {
 *   method: 'POST',
 *   body: JSON.stringify({
 *     confirmationToken: 'pi_123_secret_456',
 *     returnUrl: 'https://example.com/return'
 *   })
 * });
 * ```
 *
 * @example Client-side usage with body params
 * ```typescript
 * await fetch('/api/orders/activate-sync', {
 *   method: 'POST',
 *   body: JSON.stringify({
 *     orderId: 'ord_123',
 *     confirmationToken: 'pi_123_secret_456'
 *   })
 * });
 * ```
 */
function createActivateOrderSyncRoute(config = {}) {
    const handler = (0, billing_js_1.createActivateOrderSyncHandler)(config.defaultReturnUrl);
    const adaptedHandler = (0, base_adapter_js_1.nextjsAdapter)(handler);
    return (request, context) => adaptedHandler(request, context, config);
}
/**
 * Create a Next.js API route handler for contact creation
 *
 * @example
 * ```typescript
 * // src/app/api/contacts/route.ts
 * import { createContactsRoute } from '@paid-ai/paid-node/integrations/nextjs';
 *
 * export const POST = createContactsRoute();
 * ```
 */
function createContactsRoute(config = {}) {
    const handler = (0, contacts_js_1.createContactsHandler)();
    const adaptedHandler = (0, base_adapter_js_1.nextjsAdapter)(handler);
    return (request, context) => adaptedHandler(request, context, config);
}
/**
 * Create a Next.js API route handler for fetching customer invoices
 *
 * Organization ID is automatically fetched from the API key.
 *
 * @example
 * ```typescript
 * // src/app/api/customers/[customerExternalId]/invoices/route.ts
 * import { createCustomerInvoicesRoute } from '@paid-ai/paid-node/integrations/nextjs';
 *
 * export const GET = createCustomerInvoicesRoute();
 * ```
 */
function createCustomerInvoicesRoute(config = {}) {
    const handler = (0, invoices_js_1.createCustomerInvoicesHandler)();
    const adaptedHandler = (0, base_adapter_js_1.nextjsAdapter)(handler);
    return (request, context) => adaptedHandler(request, context, config);
}
/**
 * Create a Next.js API route handler for fetching a customer by external ID
 *
 * @example
 * ```typescript
 * // src/app/api/customers/[customerExternalId]/route.ts
 * import { createGetCustomerRoute } from '@paid-ai/paid-node/integrations/nextjs';
 *
 * export const GET = createGetCustomerRoute();
 * ```
 */
function createGetCustomerRoute(config = {}) {
    const handler = (0, customers_js_1.createGetCustomerHandler)();
    const adaptedHandler = (0, base_adapter_js_1.nextjsAdapter)(handler);
    return (request, context) => adaptedHandler(request, context, config);
}
/**
 * Create a Next.js API route handler for customer creation
 *
 * @example
 * ```typescript
 * // src/app/api/customers/route.ts
 * import { createCustomersRoute } from '@paid-ai/paid-node/integrations/nextjs';
 *
 * export const POST = createCustomersRoute();
 * ```
 *
 * @example With custom config
 * ```typescript
 * export const POST = createCustomersRoute({
 *   validate: async (body) => {
 *     if (!body.email.includes('@company.com')) {
 *       throw new Error('Only company emails allowed');
 *     }
 *   }
 * });
 * ```
 */
function createCustomersRoute(config = {}) {
    const handler = (0, customers_js_1.createCustomersHandler)();
    const adaptedHandler = (0, base_adapter_js_1.nextjsAdapter)(handler);
    return (request, context) => adaptedHandler(request, context, config);
}
/**
 * Create a Next.js API route handler for order creation
 *
 * @example
 * ```typescript
 * // src/app/api/orders/route.ts
 * import { createOrdersRoute } from '@paid-ai/paid-node/integrations/nextjs';
 *
 * export const POST = createOrdersRoute();
 * ```
 *
 * @example With auto-activation disabled
 * ```typescript
 * export const POST = createOrdersRoute({
 *   helperOptions: { autoActivate: false }
 * });
 * ```
 */
function createOrdersRoute(config = {}) {
    const handler = (0, orders_js_1.createOrdersHandler)(config.helperOptions);
    const adaptedHandler = (0, base_adapter_js_1.nextjsAdapter)(handler);
    return (request, context) => adaptedHandler(request, context, config);
}
/**
 * Create a Next.js API route handler for invoice payment
 *
 * Organization ID is automatically fetched from the API key.
 *
 * @example
 * ```typescript
 * // src/app/api/invoices/[invoiceId]/pay/route.ts
 * import { createPayInvoiceRoute } from '@paid-ai/paid-node/integrations/nextjs';
 *
 * export const POST = createPayInvoiceRoute();
 * ```
 */
function createPayInvoiceRoute(config = {}) {
    const handler = (0, billing_js_1.createPayInvoiceHandler)();
    const adaptedHandler = (0, base_adapter_js_1.nextjsAdapter)(handler);
    return (request, context) => adaptedHandler(request, context, config);
}
/**
 * Create a Next.js API route handler for user provisioning
 *
 * This creates a single endpoint that provisions a complete user
 * with customer, contact, and optional order.
 *
 * @example
 * ```typescript
 * // src/app/api/provision/route.ts
 * import { createProvisioningRoute } from '@paid-ai/paid-node/integrations/nextjs';
 *
 * export const POST = createProvisioningRoute({
 *   defaultAgentExternalId: process.env.PAID_AGENT_ID
 * });
 * ```
 *
 * @example Client-side usage
 * ```typescript
 * await fetch('/api/provision', {
 *   method: 'POST',
 *   body: JSON.stringify({ email: 'user@example.com' })
 * });
 * ```
 */
function createProvisioningRoute(config = {}) {
    const handler = (0, provision_users_js_1.createProvisioningHandler)(config.orderOptions, config.defaultAgentExternalId);
    const adaptedHandler = (0, base_adapter_js_1.nextjsAdapter)(handler);
    return (request, context) => adaptedHandler(request, context, config);
}
/**
 * Create a Next.js API route handler for setup intent creation
 *
 * Setup intents are used to add payment methods without immediate charge.
 * This is useful for saving payment methods for future billing.
 *
 * Organization ID is automatically fetched from the API key.
 *
 * @example
 * ```typescript
 * // src/app/api/payment-methods/setup/route.ts
 * import { createSetupIntentRoute } from '@paid-ai/paid-node/integrations/nextjs';
 *
 * export const POST = createSetupIntentRoute({
 *   defaultReturnUrl: process.env.NEXT_PUBLIC_BASE_URL + '/billing'
 * });
 * ```
 *
 * @example Client-side usage
 * ```typescript
 * await fetch('/api/payment-methods/setup', {
 *   method: 'POST',
 *   body: JSON.stringify({
 *     customerId: 'cus_123',
 *     confirmationToken: 'seti_123_secret_456',
 *     returnUrl: 'https://example.com/billing',
 *     metadata: { source: 'settings_page' }
 *   })
 * });
 * ```
 */
function createSetupIntentRoute(config = {}) {
    const handler = (0, billing_js_1.createSetupIntentHandler)(config.defaultReturnUrl);
    const adaptedHandler = (0, base_adapter_js_1.nextjsAdapter)(handler);
    return (request, context) => adaptedHandler(request, context, config);
}
