// @paid-ai/paid-node/dist/esm/integrations/adapters/expressjs/base-adapter.mjs
/**
 * ExpressJS adapter for Paid SDK
 * Converts Express Request/Response to framework-agnostic format
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { createFrameworkAdapter } from "../../utils/create-adapter.mjs";
/**
 * Handles Express Request/Response conversion to base handler format
 */
export const expressjsAdapter = createFrameworkAdapter({
    getBody: (req) => __awaiter(void 0, void 0, void 0, function* () {
        return req.body || {};
    }),
    getHeaders: (req) => {
        const headers = {};
        Object.entries(req.headers).forEach(([key, value]) => {
            if (typeof value === "string") {
                headers[key] = value;
            }
            else if (Array.isArray(value)) {
                headers[key] = value.join(", ");
            }
        });
        return headers;
    },
    getMethod: (req) => req.method || "GET",
    getParams: (req) => __awaiter(void 0, void 0, void 0, function* () {
        return req.params || {};
    }),
    getQuery: (req) => {
        const query = {};
        Object.entries(req.query).forEach(([key, value]) => {
            if (typeof value === "string") {
                query[key] = value;
            }
            else if (Array.isArray(value)) {
                query[key] = value.join(", ");
            }
        });
        return query;
    },
    sendJson: (data, status) => {
        throw new Error("ExpressJS adapter requires using response context from handler");
    },
});
/**
 * Create response context for ExpressJS
 * This converts the base handler response methods to ExpressJS/Express format
 */
export function createExpressJSResponseContext(res) {
    return {
        json: (data, status = 200) => {
            return res.status(status).json(data);
        },
        error: (message, status) => {
            return res.status(status).json({ error: message });
        },
    };
}
