// @paid-ai/paid-node/dist/cjs/integrations/utils/base-handler.js
"use strict";
/**
 * Framework-agnostic base handler for Paid SDK integrations
 * Provides consistent request handling, validation, and error handling
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createHandler = createHandler;
exports.createFetchHandler = createFetchHandler;
const client_js_1 = require("../client.js");
/**
 * Creates a framework-agnostic handler with automatic client initialization,
 * validation, and error handling
 *
 * @param handler - Business logic handler function
 * @param options - Handler configuration options
 * @returns Framework-agnostic handler function
 *
 * @example
 * ```typescript
 * const customersHandler = createHandler(
 *   async (client, body) => {
 *     const customer = await createCustomerWithDefaults(client, body);
 *     return { success: true, data: customer };
 *   },
 *   { requiredFields: ['externalId'] }
 * );
 * ```
 */
function createHandler(handler, options = {}) {
    const { requiredFields = [], allowedMethods = ["POST"], requireOrganizationId = false } = options;
    return function (request_1, response_1) {
        return __awaiter(this, arguments, void 0, function* (request, response, config = {}) {
            if (!allowedMethods.includes(request.method)) {
                return response.error("Method not allowed", 405);
            }
            for (const field of requiredFields) {
                if (!request.body[field]) {
                    return response.error(`${String(field)} is required`, 400);
                }
            }
            if (config.validate) {
                try {
                    yield config.validate(request.body);
                }
                catch (error) {
                    return response.error(error instanceof Error ? error.message : "Validation failed", 400);
                }
            }
            try {
                const client = (0, client_js_1.initializePaidClient)(config);
                let organizationId;
                if (requireOrganizationId) {
                    const orgId = yield (0, client_js_1.getOrganizationId)(config);
                    if (!orgId) {
                        return response.error("Organization not found", 500);
                    }
                    organizationId = orgId;
                }
                const requestData = config.transformRequest
                    ? config.transformRequest(request.body)
                    : request.body;
                const result = yield handler(client, requestData, request.params, organizationId);
                if (!result.success) {
                    return response.error(result.error || "Operation failed", result.status || 500);
                }
                const responseData = config.transformResponse
                    ? config.transformResponse(result.data)
                    : { data: result.data };
                return response.json(responseData, 200);
            }
            catch (error) {
                console.error("Handler error:", error);
                if (error instanceof TypeError && error.message.includes("fetch")) {
                    return response.error("Network error: Unable to reach Paid API", 503);
                }
                return response.error(error instanceof Error ? error.message : "Internal server error", 500);
            }
        });
    };
}
/**
 * Helper to create a fetch-based handler for simple API calls
 *
 * @param endpoint - Function that generates the endpoint URL from org ID and request body
 * @param options - Optional method, request/response transformations
 * @returns Handler function that makes the API call
 *
 * @example
 * ```typescript
 * const activateHandler = createHandler(
 *   createFetchHandler(
 *     (orgId, req) => `/api/organizations/${orgId}/orders/${req.orderId}/activate-sync`,
 *     {
 *       method: "POST",
 *       transformRequest: (req) => ({ confirmationToken: req.confirmationToken }),
 *       transformResponse: (data) => ({ order: data.order, billing: data.billing })
 *     }
 *   ),
 *   { requireOrganizationId: true }
 * );
 * ```
 */
function createFetchHandler(endpoint, options) {
    return (client, requestBody, params, organizationId) => __awaiter(this, void 0, void 0, function* () {
        var _a, _b;
        if (!organizationId) {
            return {
                success: false,
                error: "Organization ID is required for this operation",
                status: 500,
            };
        }
        try {
            const url = endpoint(organizationId, requestBody, params);
            const body = (options === null || options === void 0 ? void 0 : options.transformRequest)
                ? options.transformRequest(requestBody)
                : requestBody;
            const baseUrl = ((_a = client._options) === null || _a === void 0 ? void 0 : _a.baseUrl) || "https://api.agentpaid.io/api/v1";
            const apiKey = (_b = client._options) === null || _b === void 0 ? void 0 : _b.token;
            if (!apiKey) {
                return {
                    success: false,
                    error: "API key not found in client",
                    status: 500,
                };
            }
            const fetchResponse = yield fetch(`${baseUrl}${url}`, {
                method: (options === null || options === void 0 ? void 0 : options.method) || "POST",
                headers: {
                    Authorization: `Bearer ${apiKey}`,
                    "Content-Type": "application/json",
                },
                body: (options === null || options === void 0 ? void 0 : options.method) !== "GET" ? JSON.stringify(body) : undefined,
            });
            if (!fetchResponse.ok) {
                const errorText = yield fetchResponse.text();
                console.error("API error:", errorText);
                return {
                    success: false,
                    error: `API request failed: ${errorText}`,
                    status: fetchResponse.status,
                };
            }
            const data = yield fetchResponse.json();
            const transformedData = (options === null || options === void 0 ? void 0 : options.transformResponse)
                ? options.transformResponse(data)
                : data;
            return { success: true, data: transformedData };
        }
        catch (error) {
            console.error("Fetch error:", error);
            return {
                success: false,
                error: error instanceof Error ? error.message : "Unknown error",
                status: 500,
            };
        }
    });
}
