// @paid-ai/paid-node/dist/cjs/tracing/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.paidAutoInstrument = exports.signal = exports.trace = exports.initializeTracing = exports.getPaidTracerProvider = exports.getPaidTracer = void 0;
const tracing_js_1 = require("./tracing.js");
Object.defineProperty(exports, "getPaidTracer", { enumerable: true, get: function () { return tracing_js_1.getPaidTracer; } });
Object.defineProperty(exports, "getPaidTracerProvider", { enumerable: true, get: function () { return tracing_js_1.getPaidTracerProvider; } });
Object.defineProperty(exports, "initializeTracing", { enumerable: true, get: function () { return tracing_js_1.initializeTracing; } });
Object.defineProperty(exports, "trace", { enumerable: true, get: function () { return tracing_js_1.trace; } });
const autoInstrumentation_js_1 = require("./autoInstrumentation.js");
Object.defineProperty(exports, "paidAutoInstrument", { enumerable: true, get: function () { return autoInstrumentation_js_1.paidAutoInstrument; } });
const signal_js_1 = require("./signal.js");
Object.defineProperty(exports, "signal", { enumerable: true, get: function () { return signal_js_1.signal; } });
