// @paid-ai/paid-node/dist/esm/Client.mjs
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import * as core from "./core/index.mjs";
import { mergeHeaders } from "./core/headers.mjs";
import { Customers } from "./api/resources/customers/client/Client.mjs";
import { Agents } from "./api/resources/agents/client/Client.mjs";
import { Products } from "./api/resources/products/client/Client.mjs";
import { Contacts } from "./api/resources/contacts/client/Client.mjs";
import { Orders } from "./api/resources/orders/client/Client.mjs";
import { Plans } from "./api/resources/plans/client/Client.mjs";
import { Usage } from "./wrapper/BatchUsage.mjs";
import { Traces } from "./api/resources/traces/client/Client.mjs";
import { signal } from "./tracing/signal.mjs";
import { getPaidTracerProvider, initializeTracing, trace } from "./tracing/tracing.mjs";
export class PaidClient {
    constructor(_options = {}) {
        this._options = Object.assign(Object.assign({}, _options), { headers: mergeHeaders({
                "X-Fern-Language": "JavaScript",
                "X-Fern-SDK-Name": "",
                "X-Fern-SDK-Version": "0.8.0",
                "X-Fern-Runtime": core.RUNTIME.type,
                "X-Fern-Runtime-Version": core.RUNTIME.version,
            }, _options === null || _options === void 0 ? void 0 : _options.headers) });
    }
    get customers() {
        var _a;
        return ((_a = this._customers) !== null && _a !== void 0 ? _a : (this._customers = new Customers(this._options)));
    }
    get agents() {
        var _a;
        return ((_a = this._agents) !== null && _a !== void 0 ? _a : (this._agents = new Agents(this._options)));
    }
    get products() {
        var _a;
        return ((_a = this._products) !== null && _a !== void 0 ? _a : (this._products = new Products(this._options)));
    }
    get contacts() {
        var _a;
        return ((_a = this._contacts) !== null && _a !== void 0 ? _a : (this._contacts = new Contacts(this._options)));
    }
    get orders() {
        var _a;
        return ((_a = this._orders) !== null && _a !== void 0 ? _a : (this._orders = new Orders(this._options)));
    }
    get plans() {
        var _a;
        return ((_a = this._plans) !== null && _a !== void 0 ? _a : (this._plans = new Plans(this._options)));
    }
    get usage() {
        var _a;
        return ((_a = this._usage) !== null && _a !== void 0 ? _a : (this._usage = new Usage(this._options)));
    }
    get traces() {
        var _a;
        return ((_a = this._traces) !== null && _a !== void 0 ? _a : (this._traces = new Traces(this._options)));
    }
    /**
     * @deprecated use the standalone 'initializeTracing' instead
     */
    initializeTracing(collectorEndpoint) {
        return __awaiter(this, void 0, void 0, function* () {
            const tokenSupplier = this._options.token;
            const token = typeof tokenSupplier === "function" ? yield tokenSupplier() : tokenSupplier;
            const resolvedToken = yield Promise.resolve(token);
            initializeTracing(resolvedToken, collectorEndpoint);
        });
    }
    /**
     * @deprecated use the standalone 'trace()' instead
     * Use this method to track actions like LLM costs and sending signals.
     * The callback to this function is the work that you want to trace.
     *
     * @param externalCustomerId - The external customer ID
     * @param fn - The callback function containing the work to be traced
     * @param externalAgentId - @deprecated Use externalProductId instead
     * @param externalProductId - The external product ID
     * @param storePrompt - Whether to store the prompt (defaults to false)
     * @param args - Arguments to pass to the callback function
     * @returns The return value of the callback function
     */
    trace(externalCustomerId_1, fn_1, externalAgentId_1, externalProductId_1) {
        return __awaiter(this, arguments, void 0, function* (externalCustomerId, fn, externalAgentId, externalProductId, storePrompt = false, ...args) {
            // TODO: remove this after deprecation and replace externalAgentId with externalProductId everywhere
            if (externalProductId) {
                externalAgentId = externalProductId;
            }
            return yield trace({ externalCustomerId, externalProductId: externalAgentId, storePrompt }, fn, ...args);
        });
    }
    signal(eventName, enableCostTracingOrData, data) {
        let enableCostTracing = false;
        let finalData;
        if (typeof enableCostTracingOrData === "boolean") {
            // Case: signal(eventName, boolean, data?)
            enableCostTracing = enableCostTracingOrData;
            finalData = data;
        }
        else if (typeof enableCostTracingOrData === "object") {
            // Case: signal(eventName, data)
            enableCostTracing = false;
            finalData = enableCostTracingOrData;
        }
        // Case: signal(eventName) - both remain default/undefined
        return signal(eventName, enableCostTracing, finalData);
    }
    /**
     * @deprecated use the standalone 'getPaidTracerProvider()' instead
     * Export the tracer provider which user can use for his own tracing.
     */
    get tracerProvider() {
        return getPaidTracerProvider();
    }
}
