// @paid-ai/paid-node/dist/esm/api/resources/traces/client/Client.mjs
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import * as environments from "../../../../environments.mjs";
import * as core from "../../../../core/index.mjs";
import * as Paid from "../../../index.mjs";
import { mergeHeaders, mergeOnlyDefinedHeaders } from "../../../../core/headers.mjs";
import urlJoin from "url-join";
import * as errors from "../../../../errors/index.mjs";
export class Traces {
    constructor(_options = {}) {
        this._options = _options;
    }
    /**
     * @param {Paid.GetTracesRequest} request
     * @param {Traces.RequestOptions} requestOptions - Request-specific configuration.
     *
     * @throws {@link Paid.BadRequestError}
     * @throws {@link Paid.ForbiddenError}
     *
     * @example
     *     await client.traces.getTraces({
     *         limit: 1,
     *         offset: 1,
     *         startTime: "2024-01-15T09:30:00Z",
     *         endTime: "2024-01-15T09:30:00Z",
     *         externalCustomerId: "externalCustomerId",
     *         externalProductId: "externalProductId",
     *         externalAgentId: "externalAgentId",
     *         metadata: "metadata"
     *     })
     */
    getTraces(request = {}, requestOptions) {
        return core.HttpResponsePromise.fromPromise(this.__getTraces(request, requestOptions));
    }
    __getTraces() {
        return __awaiter(this, arguments, void 0, function* (request = {}, requestOptions) {
            var _a, _b, _c;
            const { limit, offset, startTime, endTime, externalCustomerId, externalProductId, externalAgentId, metadata } = request;
            const _queryParams = {};
            if (limit != null) {
                _queryParams["limit"] = limit.toString();
            }
            if (offset != null) {
                _queryParams["offset"] = offset.toString();
            }
            if (startTime != null) {
                _queryParams["startTime"] = startTime;
            }
            if (endTime != null) {
                _queryParams["endTime"] = endTime;
            }
            if (externalCustomerId != null) {
                _queryParams["externalCustomerId"] = externalCustomerId;
            }
            if (externalProductId != null) {
                _queryParams["externalProductId"] = externalProductId;
            }
            if (externalAgentId != null) {
                _queryParams["externalAgentId"] = externalAgentId;
            }
            if (metadata != null) {
                _queryParams["metadata"] = metadata;
            }
            const _response = yield core.fetcher({
                url: urlJoin((_b = (_a = (yield core.Supplier.get(this._options.baseUrl))) !== null && _a !== void 0 ? _a : (yield core.Supplier.get(this._options.environment))) !== null && _b !== void 0 ? _b : environments.PaidEnvironment.Production, "traces"),
                method: "GET",
                headers: mergeHeaders((_c = this._options) === null || _c === void 0 ? void 0 : _c.headers, mergeOnlyDefinedHeaders({ Authorization: yield this._getAuthorizationHeader() }), requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.headers),
                queryParameters: _queryParams,
                timeoutMs: (requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.timeoutInSeconds) != null ? requestOptions.timeoutInSeconds * 1000 : 60000,
                maxRetries: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.maxRetries,
                abortSignal: requestOptions === null || requestOptions === void 0 ? void 0 : requestOptions.abortSignal,
            });
            if (_response.ok) {
                return { data: _response.body, rawResponse: _response.rawResponse };
            }
            if (_response.error.reason === "status-code") {
                switch (_response.error.statusCode) {
                    case 400:
                        throw new Paid.BadRequestError(_response.error.body, _response.rawResponse);
                    case 403:
                        throw new Paid.ForbiddenError(_response.error.body, _response.rawResponse);
                    default:
                        throw new errors.PaidError({
                            statusCode: _response.error.statusCode,
                            body: _response.error.body,
                            rawResponse: _response.rawResponse,
                        });
                }
            }
            switch (_response.error.reason) {
                case "non-json":
                    throw new errors.PaidError({
                        statusCode: _response.error.statusCode,
                        body: _response.error.rawBody,
                        rawResponse: _response.rawResponse,
                    });
                case "timeout":
                    throw new errors.PaidTimeoutError("Timeout exceeded when calling GET /traces.");
                case "unknown":
                    throw new errors.PaidError({
                        message: _response.error.errorMessage,
                        rawResponse: _response.rawResponse,
                    });
            }
        });
    }
    _getAuthorizationHeader() {
        return __awaiter(this, void 0, void 0, function* () {
            const bearer = yield core.Supplier.get(this._options.token);
            if (bearer != null) {
                return `Bearer ${bearer}`;
            }
            return undefined;
        });
    }
}
