// @paid-ai/paid-node/dist/esm/integrations/utils/create-adapter.mjs
/**
 * Generic adapter factory - eliminates repetitive adapter code
 * Works with any framework (Express, Fastify, Hono, Next.js, etc.)
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
/**
 * Creates a framework adapter for any HTTP framework
 * Automatically extracts headers and converts request/response
 *
 * @param extractors - Framework-specific extraction functions
 * @returns Adapter function that wraps core handlers
 *
 * @example
 * ```typescript
 * const nextjsAdapter = createFrameworkAdapter({
 *   getBody: async (req) => await req.json(),
 *   getHeaders: (req) => Object.fromEntries(req.headers),
 *   getMethod: (req) => req.method || "GET",
 *   sendJson: (data, status) => NextResponse.json(data, { status })
 * });
 * ```
 */
export function createFrameworkAdapter(extractors) {
    return function adaptHandler(innerHandler) {
        return (req, context, config) => __awaiter(this, void 0, void 0, function* () {
            var _a;
            const body = yield extractors.getBody(req);
            const params = extractors.getParams
                ? yield extractors.getParams(req, context)
                : undefined;
            const requestContext = {
                body,
                headers: extractors.getHeaders(req),
                method: extractors.getMethod(req),
                params,
                query: (_a = extractors.getQuery) === null || _a === void 0 ? void 0 : _a.call(extractors, req),
            };
            const responseContext = {
                json: (data, status = 200) => extractors.sendJson(data, status),
                error: (message, status) => extractors.sendJson({ error: message }, status),
            };
            return innerHandler(requestContext, responseContext, config);
        });
    };
}
